Sources/Apollo/GraphQLQueryWatcher.swift

```diff
@@ -66,7 +69,20 @@ public final class GraphQLQueryWatcher<Query: GraphQLQuery>: Cancellable, Apollo
     guard let dependentKeys = dependentKeys else { return }
 
     if !dependentKeys.isDisjoint(with: changedKeys) {
-      fetch(cachePolicy: .returnCacheDataElseFetch)
+      // First, attempt to reload the query from the cache directly, in order not to interrupt any in-flight server-side fetch.
+      store.load(query: query) { [weak self] result in
+        guard let self = self else { return }
+
+        switch result {
+        case .success:
+          self.callbackQueue.async {
+            self.resultHandler(result)
+          }
+        case .failure:
+          // If the cache fetch is not successful, for instance if the data is missing, refresh from the server.
+          self.fetch(cachePolicy: .fetchIgnoringCacheData)
```

Comment:
is `.fetchIgnoringCacheData` correct here? it was `.returnCacheDataElseFetch` before.
Can we get into some fetch cycle here? `fetch(cachePolicy: CachePolicy)` -> Fail -> `fetch(cachePolicy: CachePolicy)` -> Fail -> ...

Reply:
Before it was `returnCacheDataElseFetch` because we weren't explicitly trying to load from the cache. By the point this gets hit, we've already checked the cache so `fetchIgnoringCacheData` acknowledges that.

Reply:
Also the `fail` is for reading directly from the store, rather than for hitting the network, so that cycle should not be an issue.